src/amber/cli/recipes/file_entries.cr

```diff
@@ -0,0 +1,69 @@
+require "teeplate"
+
+module Amber::Recipes
+  module FileEntries
+    def each_file(abs, rel, &block : String, String ->)
+      Dir.open(abs) do |d|
+        d.each_child do |entry|
+          each_file abs, rel, entry, &block
+        end
+      end
+    end
+
+    def each_file(abs, rel, entry, &block : String, String ->)
+      rel = rel ? File.join(rel, entry) : entry
+      abs = File.join(abs, entry)
+      if Dir.exists?(abs)
+        each_file abs, rel, &block
+      else
+        block.call abs, rel
+      end
+    end
+
+    def pack_liquid(files, abs, rel)
+      tpl = Liquid::Template.parse File.new(abs)
+      io = tpl.render @ctx.as(Liquid::Context)
+      files << ::Teeplate::StringData.new(rel, io.to_s, File.stat(abs).perm)
+    rescue ex
+      p "failed to process #{abs} - #{ex.message}"
+    end
+
+    def pack_blob(files, abs, rel)
+      io = IO::Memory.new
+      File.open(abs) { |f| IO.copy(f, io) }
+
+      files << ::Teeplate::Base64Data.new(rel, io.size.to_u64, Base64.encode(io), File.stat(abs).perm)
+    end
+
+    def collect_files(files)
+      if ! @template.nil?
+
+        @ctx = Liquid::Context.new
+        set_context @ctx
```

Comment:
It seems like the `set_context` method is expected to be defined in the including class. If so, what sort of error message will be rendered when a future developer adds `include FileEntries` without that method? Can it be made more helpful?

Reply:
it doesn't compile

in src/amber/cli/recipes/file_entries.cr:43: undefined method 'set_context'

        set_context @ctx